bindings/python/src/lib.rs

```diff
@@ -211,6 +211,144 @@ struct safe_open {
+            // So it's really important to make sure we are targeting the correct
+            // library.
+            unsafe {
+                let cudart = file.path();
+                let lib = libloading::Library::new(cudart).map_err(|e| {
+                    exceptions::PyException::new_err(format!("Couldn't load cuda {e:?}",))
+                })?;
+                return Ok(lib);
+            }
+        }
+    }
+    Err(exceptions::PyException::new_err("Couldn't find cuda"))
+}
+
+fn create_cuda_unsafe_tensor(
+    module: &PyModule,
+    cudart: &libloading::Library,
+    info: &TensorInfo,
+    device: &Device,
+    data: &[u8],
+) -> PyResult<PyObject> {
+    let tensor = create_empty_tensor_pt(module, &info.shape, info.dtype, device)?;
+
+    let data_ptr_fn = tensor.getattr("data_ptr")?;
+    let data_ptr: usize = data_ptr_fn.call0()?.extract()?;
+
+    // SAFETY: This is unsafe for the same reasons as when we load the library
+    let out = unsafe {
+        let cuda_memcpy: libloading::Symbol<
+            unsafe extern "C" fn(
+                device_ptr: u64,
+                src_ptr: *const std::ffi::c_void,
+                src_len: usize,
+            ) -> u32,
+        > = cudart.get(b"cudaMemcpy").map_err(|e| {
+            exceptions::PyException::new_err(format!("Couldn't find cudaMemcpy {e:?}",))
+        })?;
+        cuda_memcpy(
+            data_ptr as u64,
+            data.as_ptr() as *const std::ffi::c_void,
+            data.len(),
+        )
+    };
+    // SAFETY: Here we have a correct library, we successfully called memcpy,
+    // but somehow the call failed. This is really worrying since Pytorch is
+    // responsible for allocating the memory.
+    if out != 0 {
+        panic!(
+            "We tried to set your tensor fast, but there was a cuda error, This could
+                have corrupted your GPU ram, aborting to prevent further errors"
+        )
+    }
+    let tensor: PyObject = tensor.into_py(module.py());
+    Ok(tensor)
+}
+
+fn create_cuda_unsafe_tensor_from_slice(
+    module: &PyModule,
+    cudart: &libloading::Library,
+    shape: &[usize],
+    dtype: Dtype,
+    device: &Device,
+    iterator: safetensors::slice::SliceIterator,
+) -> PyResult<PyObject> {
+    let tensor = create_empty_tensor_pt(module, shape, dtype, device)?;
+
+    let data_ptr_fn = tensor.getattr("data_ptr")?;
+    let data_ptr: usize = data_ptr_fn.call0()?.extract()?;
+    let mut offset = 0;
+    // SAFETY: This is unsafe for the same reasons as when we load the library
+    unsafe {
+        let cuda_memcpy: libloading::Symbol<
+            unsafe extern "C" fn(
+                device_ptr: u64,
+                src_ptr: *const std::ffi::c_void,
+                src_len: usize,
+            ) -> u32,
+        > = cudart.get(b"cudaMemcpy").map_err(|e| {
+            exceptions::PyException::new_err(format!("Couldn't find cudaMemcpy {e:?}",))
+        })?;
```

Comment:
I'm not sure of the cost of that but shouldn't it be done only once ? In `__enter__` maybe

Reply:
This is a reference, and references with a lifetime and those are disallowed in `pyclass`.
There might be ways those.

Reply:
FYI I timed the actual libloading, it takes `70us` roughly..  (That's actually fetching the entire library, not just this method)

Reply:
I tried to save directly `cuda_memcpy` within the GILOnceCell, but I wasn't able to do it. Well I was able to compile it, but I triggered  the panic when using `cuda_memcpy` afterwards. So something must have been very wrong.

It was something like:

```rust
fn find_memcpy() -> Symbol<'static, MemcpyFn>{
   let symbol = Python::with_gil(|py| {
         let module = get_module(py, &TORCH_MODULE);
         let lib = find_cudart(module);
         let lib = Box::leak(Box::new(lib));
         let symbol = lib.get("cudaMemcpy");
         symbol
     })
     symbol
}
....
// At init
CUDA_MEMCPY.get_or_init(py, || find_memcpy());
```

However somehow it fails to properly call it.
`Box::leak` prevents the  `Drop` from being run right ? So the lib should not get unloaded ?

For instance:
```
The implementation of thread-local variables is extremely platform specific and uses of such variables that work on e.g. Linux may have unintended behaviour on other targets.
```
So maybe we can't really leak.

`Library` itself doesn't seem to be thread safe, maybe we should load on every access to be safe ? (It's ok speedwise)

Reply:
Hmm, that's unfortunate. It's not clear to me what might be going wrong, that looks like it should be fine to me.

I'm not aware of any thread-local variable here?

If torch is dynamically loading the cuda runtime, I wondered if it's possible that initialising too early here might interact badly with torch? I'm just speculating wildly at this point.

Reply:
If you're not worried about the loading call, you could always cache what you wanted from torch / filesystem (e.g. cuda location) and then the per-call cost is just load & lookup the symbol.

Reply:
Yes, this is what I'll end up doing I think.

Reply:
Also in `torch==1.13` they splitted cuda to it's own lib, so ....

Reply:
Ok, figure it out, loading `torch._C` directly actually works. `libloading` uses `dlopen` under the hood and is able to resolve `cudaMemcpy` directly, so no need for ELF shenanigans and it's safe since it's already loaded in the torch python namespace !